### PR TITLE
Adding Support for code based ignore elements.

### PR DESCRIPTION
**Pull request by prklm10**
Adding Support for code based ignore elements.

### CODE REVIEW

File: tests/test_cli_wrapper.py [hunk 1]
```diff
@@ -40,8 +40,20 @@ def test_request_body(self):
         tag = {'name': 'Tag'}
         name = 'some-name'
         debug_url = 'debug-url'
-        response = self.cli_wrapper._request_body(name, tag, [tile], debug_url)
+        ignored_elements_data = {
```
Comment (ninadbstack) [inline tests/test_cli_wrapper.py:43]: add a new test instead of updating existing as this is an optional param which might be null

File: percy/providers/generic_provider.py [hunk 2]
```diff
@@ -22,8 +25,9 @@ def supports(_remote_url):
     def screenshot(self, name, **kwargs):
         tiles = self._get_tiles(**kwargs)
         tag = self._get_tag(**kwargs)
+        ignore_regions = self._find_ignored_regions(**kwargs)
 
-        return self._post_screenshots(name, tag, tiles, self.get_debug_url())
+        return self._post_screenshots(name, tag, tiles, ignore_regions, self.get_debug_url())
```
Comment (ninadbstack) [inline percy/providers/generic_provider.py:30]: this is incorrect order of args, we should have a test that checks for this, clearly that test is missing

File: percy/lib/cli_wrapper.py [hunk 3]
```diff
@@ -44,8 +44,8 @@ def is_percy_enabled():
             log(e, on_debug=True)
             return False
 
-    def post_screenshots(self, name, tag, tiles, external_debug_url=None):
-        body = self._request_body(name, tag, tiles, external_debug_url)
+    def post_screenshots(self, name, tag, tiles, external_debug_url=None, ignored_elements_data=None):
```
Comment (ninadbstack) [inline percy/lib/cli_wrapper.py:47]: here last arg is ignore element data

File: percy/providers/generic_provider.py [hunk 4]
```diff
@@ -60,9 +64,91 @@ def _get_tiles(self, **kwargs):
...
+                }
+                ignored_elements_array.append(ignored_region)
+            else:
+                log(f"Values passed in custom ignored region at index: {idx} is not valid")
+
     @staticmethod
-    def _post_screenshots(name, tag, tiles, debug_url):
-        response = CLIWrapper().post_screenshots(name, tag, tiles, debug_url)
+    def _post_screenshots(name, tag, tiles, ignored_regions, debug_url):
+        response = CLIWrapper().post_screenshots(name, tag, tiles, ignored_regions, debug_url)
```
Comment (ninadbstack) [inline percy/providers/generic_provider.py:151]: here you are passing it as second last arg

File: tests/test_screenshot.py [hunk 5]
```diff
@@ -142,6 +144,32 @@ def test_posts_multiple_screenshots_to_the_local_percy_server(self):
...
+        self.mock_webdriver.find_element.return_value = mock_element
+
+        percy_screenshot(self.mock_webdriver, 'screenshot 1', ignore_regions_xpaths = xpaths)
+
+        self.assertEqual(httpretty.last_request().path, '/percy/comparison')
+
+        s1 = httpretty.latest_requests()[1].parsed_body
+
+        # self.assertEqual(s1['name'], 'screenshot 1')
+        self.assertRegex(s1['client_info'], r'percy-appium-app/\d+')
```
Comment (ninadbstack) [inline tests/test_screenshot.py:168]: where is the assertion that ignore element data is passed with expected value ?

File: tests/test_screenshot.py [hunk 6]
```diff
@@ -102,16 +104,16 @@ def test_throws_error_when_a_name_is_not_provided(self):
         with self.assertRaises(Exception):
             percy_screenshot(self.mock_webdriver)
     #
-    def test_disables_screenshots_when_the_healthcheck_fails(self):
-        mock_healthcheck(fail=True)
+    # def test_disables_screenshots_when_the_healthcheck_fails(self):
```
[the remark this answers is not in the document]
Comment (prklm10) [reply tests/test_screenshot.py:107]: locally it was giving error. forgot to uncomment it

File: tests/test_screenshot.py [hunk 7]
```diff
@@ -142,6 +144,32 @@ def test_posts_multiple_screenshots_to_the_local_percy_server(self):
...
+        mock_element.size = {"width": 200, "height": 400}
+
+        xpaths = ["//path/to/element"]
+        self.mock_webdriver.find_element.return_value = mock_element
+
+        percy_screenshot(self.mock_webdriver, 'screenshot 1', ignore_regions_xpaths = xpaths)
+
+        self.assertEqual(httpretty.last_request().path, '/percy/comparison')
+
+        s1 = httpretty.latest_requests()[1].parsed_body
```
[the remark this answers is not in the document]
Comment (prklm10) [reply tests/test_screenshot.py:165]: just copied above unit test will update both of them.

File: tests/test_cli_wrapper.py (hunk 1 above)
Comment (prklm10) [reply tests/test_cli_wrapper.py:43]: request body doesn't have optional params as the function that calls have optional params

File: percy/providers/generic_provider.py (hunk 2 above)
Comment (prklm10) [reply percy/providers/generic_provider.py:30]: it is not missing it's just that we have mocked response in test_post_screenshot so our test are running regardless.

File: tests/test_screenshot.py (hunk 5 above)
Comment (prklm10) [reply tests/test_screenshot.py]: done. it also covers the above incorrect args location.

File: percy/providers/generic_provider.py (hunk 4 above)
Comment (prklm10) [reply percy/providers/generic_provider.py]: covered in integration testing

File: tests/test_cli_wrapper.py (hunk 1 above)
Comment (ninadbstack) [reply tests/test_cli_wrapper.py]: Sure, but it validates that you are not calling a function on that arg, which can be None.

Say if your body generation included `arg.to_json()` that would break if arg was none

Comment (prklm10) [reply tests/test_cli_wrapper.py]: yep i have written new tests for it.
